dict_autotrainer.cc

```diff
@@ -0,0 +1,160 @@
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{} has no dict", s->ks_name(), s->cf_name());
+        }
+        alogger.debug("sstable_dict_autotrainer::tick(): attempting to update the dict for {}.{}", s->ks_name(), s->cf_name());
+        auto estimated_size = co_await _ss.estimate_total_sstable_volume(s->id());
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: estimated_size={}", s->ks_name(), s->cf_name(), estimated_size);
+        // constexpr uint64_t min_dataset_size = uint64_t(1)*1024*1024*1024;
+        constexpr uint64_t min_dataset_size = uint64_t(32)*1024*1024;
+        if (estimated_size < min_dataset_size) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: dataset not big enough yet (< {}), giving up on the update", s->ks_name(), s->cf_name(), min_dataset_size);
+            continue;
+        }
+        auto params = s->get_compressor_params();
+        auto ticket = get_units(_ss.get_do_sample_sstables_concurrency_limiter(), 1);
+        auto n_chunks = std::min<uint64_t>(4096, 128*1024*1024 / params.chunk_length());
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: sampling for training with chunk_length={}, n_chunks={}", s->ks_name(), s->cf_name(), params.chunk_length(), n_chunks);
+        auto training_sample = co_await _ss.do_sample_sstables(s->id(), params.chunk_length(), n_chunks);
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: got training sample with {} chunks", s->ks_name(), s->cf_name(), training_sample.size());
+        if (training_sample.size() < n_chunks) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: not enough chunks, giving up on the update", s->ks_name(), s->cf_name());
+            continue;
+        }
+        auto dict = co_await _ss.train_dict(std::move(training_sample));
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: trained dict of size {}", s->ks_name(), s->cf_name(), dict.size());
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: sampling for validation with chunk_length={}, n_chunks={}", s->ks_name(), s->cf_name(), params.chunk_length(), n_chunks);
+        auto validation_sample = co_await _ss.do_sample_sstables(s->id(), params.chunk_length(), n_chunks);
+        alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: got validation sample with {} chunks", s->ks_name(), s->cf_name(), validation_sample.size());
+        if (validation_sample.size() < n_chunks) {
+            alogger.debug("sstable_dict_autotrainer::tick(): {}.{}: not enough chunks, giving up on the update", s->ks_name(), s->cf_name());
+            continue;
+        }
+        auto ratio_before = co_await try_one_compression_config(_ss.get_database().get_user_sstables_manager().get_compressor_factory(), s, params, validation_sample);
+        auto ratio_after = co_await try_one_compression_config(dict, s, params, validation_sample);
+        constexpr float threshold_fraction = 0.95;
```

Comment:
this is another place where it's worth a comment to explain why we should publish a new dict that was estimated to be 5% worse than current one. Is it because we think it might yield a better output than current one and that the sample is not sufficient for saying its actual efficiency? Is there any math backing up this decision?

Reply:
True, I'll add a comment to this constant.

> Is it because we think it might yield a better output than current one and that the sample is not sufficient for saying its actual efficiency?

It's because we don't want to train new dictionaries (to avoid general noise on all levels) if they wouldn't be *significantly* better than the current one. So you need some arbitrary criterion for *significantly better*. And the arbitrary criterion here is that the expected compression ratio would decrease by 5%. There's no math behind this constant, and correctness isn't supposed to depend on this constant.

Reply:
Added the comments.